### ISSUE TITLE
Using urql to initialize state for editing

### ISSUE THREAD

**Issue by TriStarGod** [FormidableLabs/urql#339]
I've seen examples how to show data via query and add / delete data via mutations. However, I don't see how to edit data. In my case, I have a state that I need to initialize with the query's data. My guess is to use useEffect but I don't know how to cancel the request if the page no longer available. This is especially important if I useMutation for initialization.

for queries (not sure if an abort is needed)
```
const [result] = useQuery({ query });
useEffect(() => {
  const {fetching, data, errors} = result;
  if(!fetching) {
    if(errors) {
      setErrors(errors);
      return;
    }
  setInputs(data);
  return .... how to abort query????
},[result, setErrors, setInputs]);
```

for mutations (unsure how to abort)
```
const [res, execute] = useMutation(mutation);
useEffect(() => {
  const Init = async () => {
    const {data, error} = await execute({ input });
    if(error) {
      setErrors(error.graphQLErrors);
      return;
    }
    setInputs(data);
  }
  Init();
}, [execute, setInputs, setErrors]);
```

I'm pretty sure I'm doing the mutation part wrong but I can't think of an alternative way to initialize with a mutation. Also, why are errors for queries and mutations found in different locations?

**Comment by kitten** [comment]
I'm not quite sure what the question here is per se, but it's a usage question that's more related to React patterns on their own, right?

> I have a state that I need to initialize with the query's data

That part seems to be fine in your case, but in your code snippet you're simply copying the state, which seems odd, why aren't you just passing the query's data down into a new component, to simply state changes?

> for mutations (unsure how to abort)

Mutations are not abortable, since they end up using normal promises, which are not cancellable. You _could_ reach into the `urql` internals to get them to be canceallable, but there's little point to it, since it'd be unsafe to assume that mutations _may not happen_. I think the pattern you may want is to assume that mutations do happen and are fire-and-forget. Essentially you may want to not update too much state depending on them, especially since in normal urql usage you should assume that the query fetches again once the mutation completes.

So in summary or tl;dr: You should ensure that any state changes (from `useQuery` or otherwise) can happen safely and separate the query/mutation from the state itself to make this easier. If you assume that your "editor" (If I'm making the right assumptions about your app?) has some internal state, which can diverge from the server, you need to think about what you need to do when the data from the server changes.

This is kind of broad, I know, but this isn't an easy problem or specific to `urql`. So I'll close this for now but happy to answer more questions :)

**Comment by TriStarGod** [comment]
@kitten I see. I should pass the queried / mutation result into a new component that would manage the state. I was running the query / mutation within the same component I'm managing the state.

However, does the mutation cancel itself like the query if the component is not fully loaded? There are times I want to log when certain components are accessed and I run a mutation to initialize the component to do so. If not, I could easily move the server code to fall under type Query but I'don't believe it will follow GraphQL standards.

Also, why are errors for queries and mutations found in different locations (errors vs error.graphQLErrors)?

**Comment by kitten** [comment]
> However, does the mutation cancel itself like the query if the component is not fully loaded? 

No, mutations by default as said are not cancelled automatically and there's currently no way to do so manually without reaching into the internals, because it doesn't fit the usual GraphQL mutation patterns and practices.

> If not, I could easily move the server code to fall under type Query but I'don't believe it will follow GraphQL standards

That's true, it may not be a good idea to do any kind of database mutations in a query. However, wouldn't it be easier to do this on-demand when you're presumably sending your first save mutation?

> Also, why are errors for queries and mutations found in different locations (errors vs error.graphQLErrors)?

They're highly consistent, but the `error` you're seeing is always an instance of `CombinedError`. There's more information in the docs on what it is, but essentially it wraps any possible error that you may have and this keeps them in separate locations. it also normalises GraphQLErrors :) https://github.com/FormidableLabs/urql/blob/master/src/utils/error.ts#L41-L47
